groups/tests/test_views.py

```diff
@@ -44,6 +45,59 @@ def test_get(self):
         self.assertEqual(group, detail_object)
 
 
+class TestGroupSubscribe(RequestTestCase):
+    def setUp(self):
+        self.view = views.GroupSubscribe.as_view()
+
+    def test_form_subscribe(self):
+        group = factories.GroupFactory.create()
+        request = self.create_request('post', data={'subscribe': True})
+
+        response = self.view(request, pk=group.pk)
+
+        self.assertEqual(response.status_code, 302)
+        self.assertIn(request.user, group.watchers.all())
+        self.assertIn(group, request.user.watched_groups.all())
+
+    def test_form_unsubscribe(self):
+        group = factories.GroupFactory.create()
+        request = self.create_request('post', data={'subscribe': False})
+        group.watchers.add(request.user)
+
+        response = self.view(request, pk=group.pk)
+
+        self.assertEqual(response.status_code, 302)
+        self.assertNotIn(request.user, group.watchers.all())
+        self.assertNotIn(group, request.user.watched_groups.all())
+
+
+class TestGroupSubscribeIntegration(WebTest):
+    def setUp(self):
+        self.user = factories.UserFactory.create()
+
+    def test_subscribe_unsubscribe(self):
+        group = factories.GroupFactory.create()
+        group_url = reverse('group-detail', kwargs={'pk': group.pk})
+
+        form = self.app.get(group_url, user=self.user).form
+
+        submit_button = form.fields['subscribe-submit'][0]
+        self.assertEqual(submit_button._value, 'Subscribe')
+
+        form.submit()
+
+        self.assertIn(self.user, group.watchers.all())
+
+        form = self.app.get(group_url, user=self.user).form
+
+        submit_button = form.fields['subscribe-submit'][0]
+        self.assertEqual(submit_button._value, 'Unsubscribe')
+
+        form.submit()
+
+        self.assertNotIn(self.user, group.watchers.all())
```

Comment:
I think this test could use organising into some more logical blocks - at the moment, it's something like 40% blank lines.  Although it's not too difficult to follow, a comment or two to break up the sections wouldn't hurt either.  I'd personally go with:

```
    def test_subscribe_unsubscribe(self):
        """Check that the button performs both subscribing and unsubscribing."""
        group = factories.GroupFactory.create()
        group_url = reverse('group-detail', kwargs={'pk': group.pk})

        form = self.app.get(group_url, user=self.user).form
        submit_button = form.fields['subscribe-submit'][0]
        self.assertEqual(submit_button._value, 'Subscribe')

        # Submitting the form subscribes the user.
        form.submit()
        self.assertIn(self.user, group.watchers.all())

        # The button now says 'Unsubscribe'.
        form = self.app.get(group_url, user=self.user).form
        submit_button = form.fields['subscribe-submit'][0]
        self.assertEqual(submit_button._value, 'Unsubscribe')

        # And unsubscribes the user when clicked.
        form.submit()
        self.assertNotIn(self.user, group.watchers.all())